model/task_queue_service.go

```diff
@@ -0,0 +1,197 @@
+package model
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/evergreen-ci/evergreen/util"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+)
+
+// TaskDispatchService is an in-memory representation of schedulable tasks for a distro.
+//
+// TODO Pass all task group tasks, not just dispatchable ones, to the constructor.
+// TODO Maintain an in-memory global (but per app server) task queue per distro.
+type TaskDispatchService struct {
+	mu       sync.Mutex
```

Comment:
rw mutex pls

Reply:
Only `FindNextTask()` is mutexed, and it is mutexed for both reading and writing.

Reply:
is that enough?

Reply:
it should be unless we add read-only methods on it